### PR TITLE
Reduce output buffer lock contention

### PR DESCRIPTION
**Pull request by pettyjamesm**
Two commits attempting to reduce lock contention related to output buffers:
- Refactors `LazyOutputBuffer` to attempt to read the delegate `OutputBuffer` reference before synchronizing
- Refactors `SerializedPageReference` and usages to perform dereferencing in batches to avoid synchronizing on memory manager locks for each individual page.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by pettyjamesm** [comment]
Converted to draft after identifying another very significant source of related contention while I test out a possible solution.

**Review by rschlussel — COMMENTED** [review]
"Avoid unnecessary synchronization in LazyOutputBuffer" generally looks good.  Makes the code a bit more complex though.  Do you have data about the performance improvements from this change?

**Comment by pettyjamesm** [comment]
> "Avoid unnecessary synchronization in LazyOutputBuffer" generally looks good. Makes the code a bit more complex though. Do you have data about the performance improvements from this change?

The changes in that particular commit don't improve things measurably on their own, although they do avoid a thread-serializing effect for accesses to the delegate `OutputBuffer` through the `LazyOutputBuffer` in the common case. The bigger improvement is in the second commit which avoids contending for the root contended lock (the memory tracking system) on a per-page basis by batching those operations instead.

**Review by rschlussel — COMMENTED** [review]
I really like how this change also allows a nice refactoring for the output buffers to share a page tracking implementation.

One question I have is, is there a chance that by batching the page dereferences, we could end up holding on to some memory for too long after we're done with it? 

Otherwise looks good.

@wenleix it would be great if you could take a look at the second commit, since you have worked with the output buffer/page refernce code before.  (main classes are SerializedPageReference and LifespanSerializedPageTracker, and the different output buffers are updated to use these)

**Comment by pettyjamesm** [comment]
> One question I have is, is there a chance that by batching the page dereferences, we could end up holding on to some memory for too long after we're done with it?

I don’t think so (at least in this PR) because no significant work is performed or locks acquired (except for the target lock of the memory tracking system) between pages under the per-page dereferencing mode. This should just turn N rapid lock acquires/releases into 1 (assuming the pages share a lifespan, although even if they don’t the degradation mode returns to N) which would never be worse than the worst case before but is better than the previous common case. The “extra time” will be the duration of a summing-loop over a smallish list.

**Review by rschlussel — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java:96]: nit: we can use `decrementAndGet(releasedPageCount)`

File: presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java:103]: I didn't see this `childListener` related logic in https://github.com/trinodb/trino/pull/6097/ ? And looks like it's only used in `BroadcastOutputBuffer` as a safety check: 

```java
checkState(totalBufferedPages.addAndGet(-releasedPageCount) >= 0);
```

File: presto-main/src/main/java/com/facebook/presto/execution/buffer/BroadcastOutputBuffer.java
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/buffer/BroadcastOutputBuffer.java:91]: this check looks quite universal -- wondering why it cannot be used for other buffer as well?

File: presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java:65]: isn't `computeIfAbsent` will return the counter in the map if exist? -- so we don't need this extra get and can directly do
```java
counter = outstandingPageCountPerLifespan.computeIfAbsent(lifespan, ignore -> new AtomicLong())
```

**Review by wenleix — COMMENTED** [review]
Lifespan related changed LGTM % minor comments.

File: presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java
Comment (pettyjamesm) [reply presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java:65]: In JDK8 `ConcurrentHashMap#computeIfAbsent` unconditionally acquires a write lock even when the value is present (and even so, would unconditionally create lambda in any JDK version). Later JDK's address that, but since prestodb still supports JDK8 I chose to go this route.

File: presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java
Comment (pettyjamesm) [reply presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java:96]: `decrementAndGet()` takes no arguments as it implies a value of `-1`

File: presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java
Comment (pettyjamesm) [reply presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java:103]: That's correct. Since trino doesn't perform that sanity check there wasn't a need to add the same sub-listener logic. I'm not sure you'd want to add another atomic counter to the other buffers as they all have a more straightforward way of counting currently buffered pages (eg: `bufferedPages.size()`)

File: presto-main/src/main/java/com/facebook/presto/execution/buffer/BroadcastOutputBuffer.java
Comment (pettyjamesm) [reply presto-main/src/main/java/com/facebook/presto/execution/buffer/BroadcastOutputBuffer.java:91]: See above comments, but I'm not sure the extra cost of tracking the count makes sense there since the number of buffered pages is more easily reasoned about through other means except for the broadcast buffer.

File: presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java:65]: @pettyjamesm : I see. In that case add a comment explain why not directly using `computeIfAbsent` patern. Also since line 67 is not the common path, shall we just use `putIfAbsent` on line 67 instead?

**Review by wenleix — COMMENTED** [review]
Lifespan part of change LGTM % one only minor comment.

File: presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java
Comment (pettyjamesm) [reply presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java:66]: `putIfAbsent` unfortunately returns the _previous_ value which makes it more awkward to use in this context than `computeIfAbsent`. I've added a comment about the common case avoiding a lambda creation but omitted the minutia about `ConcurrentHashMap` locking behaviors in JDK 8 since people won't care about that once it's no longer supported.

Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java:66]: @pettyjamesm : The major motivation of the comment is so maintainers of the code can refactor the code later by reading the comment, what about some comment like following:

```
/  In JDK8 ConcurrentHashMap#computeIfAbsent unconditionally acquires a write lock even when the value is present
// TODO: Remove this get ConcurrentHashMap#get call once Presto no longer supports JDK-8
```

**Review by wenleix — APPROVED** [review]

**Comment by wenleix** [comment]
Thanks. I have no further comments. Will merge the PR once Travis is clean.

**Pull request merged**
